module-services/service-cellular/Pdu.cpp

```diff
@@ -0,0 +1,978 @@
+            for (std::uint32_t i = 0; i < length; ++i) {
+                std::uint8_t octet = 0x00;
+                if (!octets.next(octet)) {
+                    return false;
+                }
+                val = (val << 8) | octet;
+            }
+            return true;
+        }
+
+        // http://www.openmobilealliance.org/release/Browser_Protocol_Stack/V2_1-20110315-A/OMA-WAP-TS-WSP-V1_0-20110315-A.pdf,
+        // p79
+        static bool parseTextN(Octets &octets, Text &text, std::uint32_t length)
+        {
+            if (length > StringMaxLength) {
+                return false;
+            }
+            if (length == 0) {
+                text = Text(octets.begin(), octets.begin());
+                return true;
+            }
+            // Avoid including EoS character
+            Octets::Iterator const begin = octets.begin();
+            octets.eatN(length - 1);
+            Octets::Iterator const end = octets.begin();
+            std::uint8_t octet         = 0x00;
+            if (!octets.next(octet)) {
+                return false;
+            }
+            if (octet == EndOfString) {
+                text = Text(begin, end);
+            }
+            else {
+                text = Text(begin, octets.begin());
+            }
+            return true;
+        }
+
+        // http://www.openmobilealliance.org/release/Browser_Protocol_Stack/V2_1-20110315-A/OMA-WAP-TS-WSP-V1_0-20110315-A.pdf,
+        // p79
+        static bool parseText(Octets &octets, Text &text)
+        {
+            Octets::Iterator begin = octets.begin();
+            std::uint8_t octet     = 0x00;
+            if (!octets.next(octet)) {
+                return false;
+            }
+            if (octet == TextEscape) {
+                begin = octets.begin();
+            }
+            // Avoid including EoS character
+            Octets::Iterator end = begin;
+            while (octet != EndOfString) {
+                end = octets.begin();
+                if (!octets.next(octet)) {
+                    return false;
+                }
+            }
+            text = Text(begin, end);
+            return true;
+        }
+
+        // http://www.openmobilealliance.org/release/Browser_Protocol_Stack/V2_1-20110315-A/OMA-WAP-TS-WSP-V1_0-20110315-A.pdf,
+        // p79
+        // Value may be:
+        // - decoded short integer
+        // - text
+        // - data as parsable octets range
+        struct DefaultPolicy
+        {
+            static bool data(Octets &octets, std::uint8_t lengthOctet, DefaultValue &value)
+            {
+                octets.eat();
+                std::uint32_t length = lengthOctet;
+                if (lengthOctet == UintvarEscape) {
+                    if (!parseUintvar(octets, length)) {
+                        return false;
+                    }
+                }
+                if (octets.size() < length) {
+                    return false;
+                }
```

Comment:
I've added this check and similar one below to return 'false' if an incomplete message was sent. Or several partial messages were not correctly concatenated into one notification. Before this change the returned value contained as many characters as possible and parsing was considered as successful.
There is an edge case however. It's impossible to detect incomplete message if it ends exactly where one of the headers ends. In such case `true` will still be returned and incomplete number of headers will be parsed.